cmd/influx_inspect/cardinality/aggregators.go

```diff
@@ -0,0 +1,173 @@
+	for _, s := range headers {
+		div = append(div, strings.Repeat("-", len(s)))
+	}
+	return strings.Join(div, "\t")
+}
+
+func newSimpleMeasurementFactory(newCounterFn func() report.Counter, est string) *measurementFactory {
+	return &measurementFactory{
+		header:   simpleHeader,
+		estTitle: est,
+		newNode:  func(inner bool) node { return newSimpleNode(inner, newCounterFn) },
+		counter:  newCounterFn,
+	}
+}
+
+func newDetailedMeasurementFactory(newCounterFn func() report.Counter, est string) *measurementFactory {
+	return &measurementFactory{
+		header:   detailedHeader,
+		estTitle: est,
+		newNode:  func(inner bool) node { return newDetailedNode(inner, newCounterFn) },
+		counter:  newCounterFn,
+	}
+}
+
+type simpleNode struct {
+	sync.Mutex
+	report.Counter
+	nodeMap
+}
+
+func (s *simpleNode) nextLevel() nodeMap {
+	return s.nodeMap
+}
+
+func newSimpleNode(inner bool, fn func() report.Counter) *simpleNode {
+	s := &simpleNode{Counter: fn()}
+	if inner {
+		s.nodeMap = make(nodeMap)
+	}
+	return s
+}
+
+func (s *simpleNode) recordMeasurement(dbRp string, key, field []byte, tags models.Tags, newCounterFn func() report.Counter) {
+	s.Add(key)
+}
+
+func (s *simpleNode) print(tw *tabwriter.Writer, db, rp, ms string) error {
+	_, err := fmt.Fprintf(tw, "%s\t%s\t%s\t%d\n",
+		db,
+		rp,
+		ms,
+		s.Count())
+	return err
+}
+
+type detailedNode struct {
+	simpleNode
+	fields report.Counter
+	tags   map[string]report.Counter
+}
+
+func (d *detailedNode) nextLevel() nodeMap {
+	return d.nodeMap
+}
+
+func newDetailedNode(inner bool, fn func() report.Counter) *detailedNode {
+	d := &detailedNode{
+		simpleNode: simpleNode{
+			Counter: fn(),
+		},
+		fields: fn(),
+		tags:   make(map[string]report.Counter),
+	}
+	if inner {
+		d.simpleNode.nodeMap = make(nodeMap)
+	}
+	return d
+}
+
+func (d *detailedNode) recordMeasurement(dbRp string, key, field []byte, tags models.Tags, newCounterFn func() report.Counter) {
+	d.Add(key)
+	d.fields.Add(field)
+	for _, t := range tags {
+		// Add database and retention policy to correctly aggregate
+		// in inner (non-leaf) nodes
+		canonTag := dbRp + string(t.Key)
+		tc, ok := d.tags[canonTag]
+		if !ok {
+			tc = newCounterFn()
+			d.tags[canonTag] = tc
+		}
+		tc.Add(t.Value)
+	}
+}
+
+func (d *detailedNode) print(tw *tabwriter.Writer, db, rp, ms string) error {
+	seriesN := d.Count()
+	fieldsN := d.fields.Count()
+	var tagKeys []string
+	tagN := uint64(0)
+
+	if d.nodeMap == nil {
+		tagKeys = make([]string, 0, len(d.tags))
+	}
+	for k, v := range d.tags {
+		c := v.Count()
+		tagN += c
+		if d.nodeMap == nil {
+			tagKeys = append(tagKeys, fmt.Sprintf("%q: %d", k, c))
+		}
+	}
+	_, err := fmt.Fprintf(tw, "%s\t%s\t%s\t%d\t%d\t%d\t%s\n",
+		db,
+		rp,
+		ms,
+		seriesN,
+		fieldsN,
+		tagN,
+		strings.Join(tagKeys, ", "))
+	return err
+}
+
+func initRecord(n node, dbRp string, key []byte, field []byte, tags models.Tags, nodeFn func(bool) node, counterFn func() report.Counter, levelKeys ...string) {
```

Comment:
I think the rollup level could be encapsulated here (and written explicitly), and then instead of `levelKeys` we could write `db, rp, measurement` which would be clearer.

Especially if `initRecord` was a method on the top-level node, and the rollup depth a member there too.

Something like:

```
func (n *node) Record(db, rp, measurement string, key, field []byte, tags models.Tags) {
  dbRp := fmt.Sprintf("%s/%s", db, rp)
  n.record(dbRp, key, field, tags) // counterFn should be a member on the node I think
  if n.aggregateDepth <= 0 {
    return
  }
  dbNode = n.child(db) // child creates the new node if needed
  dbNode.record(dbRp, key, field, tags)
  if n.aggregateDepth <= 1 {
    return
  }
  rpNode = dbNode.child(rp) // child creates the new node if needed
  rpNode.record(dbRp, key, field, tags)
  if n.aggregateDepth <= 2 {
    return
  }
  measurementNode = rpNode.child(measurement) // child creates the new node if needed
  measurementNode.record(dbRp, key, field, tags)
}
```

Reply:
Calling the top level node `measurementNode` was confusing.

Reply:
It would also be helpful to extract this to a different module I think (maybe an `internal` module if we're not sharing it), to distinguish between internal helper methods and external contract methods.

Reply:
Fixing some naming.  Calling the interface `rollupNode`, creating `isLeaf()` and `children()` methods, etc.  Will look into your    suggestion of making `initRecord` a method, too.